prerendering-state.md

```diff
@@ -0,0 +1,150 @@
+# Prerendering State API
+
+## Summary
+
+We propose introducing a boolean property `document.prerendering` and associated change event to distinguish
+prerendering browsing contexts from regular ones.
+
+_Note: This is based on the more general discussion in this [background
+doc](https://docs.google.com/document/d/1Xzw0k8DgltI2ohapuDKmjRZLv7NVrRFGusW8IBtiCT0/edit?usp=sharing)_
+
+## Background
+
+Pages need to know when they're being rendered inside a [prerendering browsing context](browsing-context.md).
+Some examples of divergent behavior:
+
+* Avoid fetching large video resources
+* Avoid measuring ad impressions
+* Avoid running javascript animations
+* Avoid measuring user interactions for analytics
+* Avoid starting server-coordinated behavior (e.g. ticket purchase countdown).
+
+A previous iteration of prerendering used [page-visibility](https://www.w3.org/TR/page-visibility-2) to denote a
+prerendering context. However, this was unimplemented and [removed from the
+specification](https://github.com/w3c/page-visibility/issues/42).
+
+With the re-introduction of prererendering, we need to bring back some form of this API.
+
+## Why not visibilityState=='prerender'?
+
+The most straightforward way to do this would be to bring back the previously specified `visibilityState ==
+'prerender'`.  Doing so has a number of drawbacks:
+
+### Not mutually-exclusive
+
+In actuality, some of the above use cases are interested in visibility rather than prerendering state.  For example, an
+ordinary (i.e. non-prerendering) page that's hidden would likely want to "avoid running javascript animations" and
+"avoid fetching large video resources". Thus, pages would have to consider both the `hidden` state as well as a
+re-introduced `prerender` to get the desired behavior.
+
+However, the [portals proposal](https://github.com/WICG/portals/) means that `prerender` isn't mutually exclusive
+with `visible` either. In that case, we do want to "run javascript animations" and "fetch large resources" since it'll
+be visible to the user, despite being in a prerendering browsing context (and all the restrictions that come with that).
+
+### Web-compatibility
+
+Related to the above point, a common pattern on the web is to assume only two visibility states. For example, code such
+as:
+
+```
+onload = () => {
+  if (document.visibilityState == 'visible')
+    // Load large images
+  else
+    // Defer loading large images
+}
+```
+
+An analysis [found several
+examples](https://docs.google.com/document/d/1Xzw0k8DgltI2ohapuDKmjRZLv7NVrRFGusW8IBtiCT0/edit#heading=h.rkorueof7xev)
+of such behavior.
+
+(re-)Introducing a `prerender` state could break such pages. In particular, portals (or some other, future
+prerendering-while-visible context) would show static or missing content. A hidden prerender may not have immediate
```

Comment:
The term "preview" might be a more intuitive way of describing prerendering-while-visible.

Reply:
I agree, but I think it makes sense for this be a signal that we're in a prerendering-browsing-context. If that's the case, I think it'd make sense to match the names of the two things;  "prerendering-browsing-context" has the same issue. Would it make sense to rename them both? @jeremyroman @domenic for their thoughts.

Reply:
Ah, I was only suggesting we might use the term preview in the prose describing what the new prerender state did, not changing the name of anything web-facing.

Reply:
Ah - true, but that does beg the question whether naming the web-facing APIs 'prerender' evokes the right meaning if we need to clarify in prose

Reply:
This is still outstanding but unresolved. I filed #12 to keep this in mind but I think we can punt on this for now and leave it as "prerendering" since it's much clearer to outsiders what we're talking about. As other use cases shape up we can revisit it in the future.